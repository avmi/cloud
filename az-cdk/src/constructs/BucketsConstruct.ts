import { Construct, RemovalPolicy } from '@aws-cdk/core';
import { Bucket, HttpMethods, CorsRule } from '@aws-cdk/aws-s3';

export interface IBucketParams {
  name: string;
  corsGET?: boolean;
  corsPUT?: boolean;
  corsPOST?: boolean;
  corsDELETE?: boolean;
}

export interface IBucketsProps {
  buckets: IBucketParams[];
}

export class BucketsConstruct extends Construct {
  readonly buckets: Bucket[] = [];

  constructor(scope: Construct, id: string, props: IBucketsProps) {
    super(scope, id);

    props.buckets.forEach((b, i) => {
      let cors: CorsRule[] | undefined = undefined;
      if (b.corsGET || b.corsPUT || b.corsPOST || b.corsDELETE) {
        const allowedMethods: HttpMethods[] = [];
        if (b.corsGET) {
          allowedMethods.push(HttpMethods.GET);
        }
        if (b.corsPUT) {
          allowedMethods.push(HttpMethods.PUT);
        }
        if (b.corsPOST) {
          allowedMethods.push(HttpMethods.POST);
        }
        if (b.corsDELETE) {
          allowedMethods.push(HttpMethods.DELETE);
        }
        cors = [
          {
            allowedOrigins: ['*'],
            allowedHeaders: ['*'],
            allowedMethods,
          },
        ];
      }

      const bucket = new Bucket(this, `Bucket${i}`, {
        bucketName: b.name,
        removalPolicy: RemovalPolicy.DESTROY,
        cors,
      });

      this.buckets.push(bucket);
    });
  }
}
