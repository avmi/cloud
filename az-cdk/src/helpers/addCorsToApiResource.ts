import { IResource, MockIntegration, PassthroughBehavior } from '@aws-cdk/aws-apigateway';

export const addCorsToApiResource = (
  resource: IResource,
  methods: string[] = ['GET', 'PUT', 'POST', 'DELETE'],
) => {
  const m = methods.join(',');
  resource.addMethod(
    'OPTIONS',
    new MockIntegration({
      integrationResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Headers': `'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'`,
            'method.response.header.Access-Control-Allow-Origin': `'*'`,
            'method.response.header.Access-Control-Allow-Credentials': `'false'`,
            'method.response.header.Access-Control-Allow-Methods': `'OPTIONS,${m}'`,
          },
        },
      ],
      passthroughBehavior: PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': '{"statusCode": 200}',
      },
    }),
    {
      methodResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Headers': true,
            'method.response.header.Access-Control-Allow-Methods': true,
            'method.response.header.Access-Control-Allow-Credentials': true,
            'method.response.header.Access-Control-Allow-Origin': true,
          },
        },
      ],
    },
  );
};
