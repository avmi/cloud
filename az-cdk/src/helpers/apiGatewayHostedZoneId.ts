/*
  The API Gateway HostedZoneId for each region has to be looked up here:
     https://docs.aws.amazon.com/general/latest/gr/rande.html#apigateway_region

  Examples:
    US East (N. Virginia)      us-east-1       execute-api.us-east-1.amazonaws.com       Z1UJRXOUMOOFQ8
    US West (N. California)    us-west-1       execute-api.us-west-1.amazonaws.com       Z2MUQ32089INYE
    US East (Ohio)             us-east-2       execute-api.us-east-2.amazonaws.com       ZOJJZC49E0EPZ
    US West (Oregon)           us-west-2       execute-api.us-west-2.amazonaws.com       Z2OJLYMUO9EFXC
    Asia Pacific (Hong Kong)   ap-east-1       execute-api.ap-east-1.amazonaws.com       Z3FD1VL90ND7K5
    Asia Pacific (Singapore)   ap-southeast-1  execute-api.ap-southeast-1.amazonaws.com  ZL327KTPIQFUL
    Asia Pacific (Sydney)      ap-southeast-2  execute-api.ap-southeast-2.amazonaws.com  Z2RPCDW04V8134
    South America (Sao Paulo)  sa-east-1       execute-api.sa-east-1.amazonaws.com       ZCMLWB8V5SYIT
*/

export const apiGatewayHostedZoneId = (region: string) => {
  switch (region.toLowerCase()) {
    case 'us-east-2':
      return 'ZOJJZC49E0EPZ';
    case 'us-east-1':
      return 'Z1UJRXOUMOOFQ8';
    case 'us-west-1':
      return 'Z2MUQ32089INYE';
    case 'us-west-2':
      return 'Z2OJLYMUO9EFXC';
    case 'ap-east-1':
      return 'Z3FD1VL90ND7K5';
    case 'ap-south-1':
      return 'Z3VO1THU9YC4UR';
    case 'ap-northeast-2':
      return 'Z20JF4UZKIW1U8';
    case 'ap-southeast-1':
      return 'ZL327KTPIQFUL';
    case 'ap-southeast-2':
      return 'Z2RPCDW04V8134';
    case 'ap-northeast-1':
      return 'Z1YSHQZHG15GKL';
    case 'ca-central-1':
      return 'Z19DQILCV0OWEC';
    case 'eu-central-1':
      return 'Z1U9ULNL0V5AJ3';
    case 'eu-west-1':
      return 'ZLY8HYME6SFDD';
    case 'eu-west-2':
      return 'ZJ5UAJN8Y3Z2Q';
    case 'eu-west-3':
      return 'Z3KY65QIEKYHQQ';
    case 'eu-north-1':
      return 'Z3UWIKFBOOGXPP';
    case 'me-south-1':
      return 'Z20ZBPC0SS8806';
    case 'sa-east-1':
      return 'ZCMLWB8V5SYIT';
  }
  throw new Error(`cannot get API Gateway HostedId for region "${region}"`);
};
