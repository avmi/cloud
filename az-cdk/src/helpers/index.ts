export * from './apiGatewayHostedZoneId';
export * from './envars2cdk';
export * from './npmCommands';
export * from './ssmSecret';
export * from './synthApp';
