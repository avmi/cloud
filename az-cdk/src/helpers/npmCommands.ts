export interface INPMCommands {
  install: string[];
  build: string[];
}

export const npmCommands = (extraBeforeTest = '', useYarn = false): INPMCommands => {
  if (useYarn) {
    return {
      install: [
        'curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -',
        'echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list',
        'sudo apt-get update && sudo apt-get install yarn',
        'yarn',
      ],
      build: ['yarn test', 'yarn build'],
    };
  }
  return {
    install: ['npm install'],
    build: extraBeforeTest
      ? [extraBeforeTest, 'npm run test', 'npm run build']
      : ['npm run test', 'npm run build'],
  };
};
