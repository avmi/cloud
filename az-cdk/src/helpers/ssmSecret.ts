import { SecretValue, CfnDynamicReference, CfnDynamicReferenceService } from '@aws-cdk/core';

// ssmSecret creates a reference to a parameter in the SSM Parameter Store (not Secrets Manager!)
export const ssmSecret = (input: { name: string; version: string }): SecretValue => {
  return SecretValue.cfnDynamicReference(
    new CfnDynamicReference(CfnDynamicReferenceService.SSM, `${input.name}:${input.version}`),
  );
};
