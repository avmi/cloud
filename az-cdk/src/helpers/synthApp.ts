import { App } from '@aws-cdk/core';

// synthApp synthetizes App and returns object
export const synthApp = (app: App): any => {
  const assembly = app.synth();
  const synthesized = assembly.stacks[0];
  return synthesized.template;
};

// synthAppString synthetizes App and returns stringifyied object
export const synthAppString = (app: App): string => {
  return JSON.stringify(synthApp(app), undefined, 2);
};
