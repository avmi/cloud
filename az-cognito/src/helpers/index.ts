export * from './flattenAttributes';
export * from './email2username';
