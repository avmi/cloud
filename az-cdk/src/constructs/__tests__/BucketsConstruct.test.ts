import { App, Stack } from '@aws-cdk/core';
import { synthAppString } from '../../helpers/synthApp';
import { BucketsConstruct } from '../BucketsConstruct';

describe('BucketsConstruct ', () => {
  it('synthetizes properly', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new BucketsConstruct(stack, 'Buckets', {
      buckets: [{ name: 'balde' }, { name: 'deposito' }],
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });

  it('synthetizes properly with CORS', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new BucketsConstruct(stack, 'Buckets', {
      buckets: [
        {
          name: 'balde',
          corsDELETE: true,
          corsGET: true,
          corsPOST: true,
          corsPUT: true,
        },
        {
          name: 'deposito',
          corsPUT: true,
        },
      ],
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });
});
