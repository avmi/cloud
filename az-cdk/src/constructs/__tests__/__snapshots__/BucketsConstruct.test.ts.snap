// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BucketsConstruct  synthetizes properly 1`] = `
"{
  \\"Resources\\": {
    \\"BucketsBucket04F23AD6D\\": {
      \\"Type\\": \\"AWS::S3::Bucket\\",
      \\"Properties\\": {
        \\"BucketName\\": \\"balde\\"
      },
      \\"UpdateReplacePolicy\\": \\"Delete\\",
      \\"DeletionPolicy\\": \\"Delete\\"
    },
    \\"BucketsBucket1B98A5B53\\": {
      \\"Type\\": \\"AWS::S3::Bucket\\",
      \\"Properties\\": {
        \\"BucketName\\": \\"deposito\\"
      },
      \\"UpdateReplacePolicy\\": \\"Delete\\",
      \\"DeletionPolicy\\": \\"Delete\\"
    }
  }
}"
`;

exports[`BucketsConstruct  synthetizes properly with CORS 1`] = `
"{
  \\"Resources\\": {
    \\"BucketsBucket04F23AD6D\\": {
      \\"Type\\": \\"AWS::S3::Bucket\\",
      \\"Properties\\": {
        \\"BucketName\\": \\"balde\\",
        \\"CorsConfiguration\\": {
          \\"CorsRules\\": [
            {
              \\"AllowedHeaders\\": [
                \\"*\\"
              ],
              \\"AllowedMethods\\": [
                \\"GET\\",
                \\"PUT\\",
                \\"POST\\",
                \\"DELETE\\"
              ],
              \\"AllowedOrigins\\": [
                \\"*\\"
              ]
            }
          ]
        }
      },
      \\"UpdateReplacePolicy\\": \\"Delete\\",
      \\"DeletionPolicy\\": \\"Delete\\"
    },
    \\"BucketsBucket1B98A5B53\\": {
      \\"Type\\": \\"AWS::S3::Bucket\\",
      \\"Properties\\": {
        \\"BucketName\\": \\"deposito\\",
        \\"CorsConfiguration\\": {
          \\"CorsRules\\": [
            {
              \\"AllowedHeaders\\": [
                \\"*\\"
              ],
              \\"AllowedMethods\\": [
                \\"PUT\\"
              ],
              \\"AllowedOrigins\\": [
                \\"*\\"
              ]
            }
          ]
        }
      },
      \\"UpdateReplacePolicy\\": \\"Delete\\",
      \\"DeletionPolicy\\": \\"Delete\\"
    }
  }
}"
`;
